Refactor /decide enpoint & allow recording without autocapture

paolodamico opened the pull request:
## Changes

This PR enables running session recording even if autocapture is disabled. This is motivated by a couple of requests from different users who wish to use session recording without autocapture.
- To enable this, we refactor the `/decide` endpoint logic out of the `autocapture` module.

## Checklist
- [X] Tests for new code (if applicable)
- [X] TypeScript definitions (module.d.ts) updated and in sync with library exports. **Not applicable**

macobo commented on cypress/integration/session-recording.spec.js:
```diff
@@ -8,7 +8,7 @@ describe('Session recording', () => {
             url: '**/decide/*',
             response: {
-                config: { enable_collect_everything: true },
+                config: { enable_collect_everything: false },
```
Q: Why this change?

paolodamico replied:
Intended to make sure that session recording works even if autocapture is off

macobo commented on src/decide.js:
```diff
@@ -0,0 +1,56 @@
+import { autocapture } from './autocapture'
+import { _ } from './utils'
+
+const decide = {
```
Code style nit: Let's use classes as with `src/extensions/sessionrecording.js`.

So constrctor + `callDecide` function.

paolodamico replied:
Ack had it as a class first but thought it was more straightforward like this, reverting to class 😉

macobo commented on src/posthog-core.js:
```diff
@@ -131,6 +132,8 @@ var create_mplib = function (token, config, name) {
     instance.sessionRecording = new SessionRecording(instance)
     instance.sessionRecording.startRecordingIfEnabled()
 
+    decide.init(instance)
```
So this refactoring has a non-intuitive side-effect which might be considered a regression - we now always call /decide. 

Previously decide was not called when autocapture was off - making the script a lot more light-weight if nothing was turned on.

paolodamico replied:
Ack you're right. Well definitely something worth changing would be to not call the `/decide` endpoint if session recording and autocapture are both disabled. Perhaps if we want to avoid the extra request to `/decide` we need to introduce a different param that enables session recording even if autocapture is off? Thoughts?

paolodamico replied:
Per sync conversation, we'll keep the new approach as `/decide` is also required for other stuff like compression and feature flags, and adding an advanced config option to fully disable this request.

macobo commented on src/decide.js:
```diff
@@ -0,0 +1,56 @@
+            instance.get_config('api_host') + '/decide/',
+            { data: encoded_data },
+            { method: 'POST' },
+            instance._prepare_callback(this.parseDecideResponse)
```
Note: `_prepare_callback` is not really needed. posthog-js does not really work unless you use xhr.

macobo commented on src/decide.js:
```diff
@@ -0,0 +1,56 @@
+            distinct_id: instance.get_distinct_id(),
+        })
+
+        var encoded_data = _.base64Encode(json_data)
```
es6 please (var => const)!

macobo commented on src/decide.js:
```diff
@@ -0,0 +1,56 @@
+        )
+    },
+
+    parseDecideResponse(response) {
```
This is missing tests - should we add them?

paolodamico replied:
Done!

macobo reviewed (commented):
Note - the idea seems good, one worry about a regression below.

However from a product perspective is this a case of perverse incentives since we do not bill session recordings? Load and cost-wise session recordings >>>> autocapture for us.

paolodamico commented:
I've updated the PR based on what we talked about and is ready for a review. Will create a complimentary PR to update the relevant docs. Not sure why the testcafe tests are failing, can you give me some pointers on how to debug? Can't seem to find any particularly useful information on Browserstack.

macobo commented on src/__tests__/compression.js:
```diff
@@ -98,6 +99,8 @@ describe('Payload Compression', () => {
         })
 
         it('should save supported compression in instance', () => {
+            const decide = new Decide(lib)
```
Note: you can shorten this to `new Decide(lib).call()`

macobo commented on src/__tests__/decide.js:
```diff
@@ -0,0 +1,90 @@
+
+    given('config', () => ({ api_host: 'https://test.com' }))
+
+    jest.spyOn(autocapture, 'afterDecideResponse').mockImplementation()
```
This should be in a beforeEach block.

macobo commented on src/posthog-core.js:
```diff
@@ -722,7 +732,13 @@ PostHogLib.prototype._register_single = function (prop, value) {
  */
 PostHogLib.prototype.isFeatureEnabled = function (key, options = {}) {
-    return this.feature_flags.isFeatureEnabled(key, options)
+    if (!this.featureFlags) {
```
This change has a couple of issues:

1. Returning undefined makes no sense and can break peoples code. This function should return a boolean. `false` is the good default
2. Checking `!this.featureFlags` is misleading as it will fail if no flags are turned on.
3. This will cause logspam - this is always reached if isFeatureEnabled is called before decide returns a response.
4. This doesn't belong to posthog-core.js at all - feature flags file should handle all the logic here. Otherwise the two ways to call featureflags in the library diverge in the implementation.

I suggest not including this change at all.

paolodamico replied:
Removed this code, users will get an error if attempting to load feature flags with `/decide` endpoint disabled anyways.

macobo commented on src/posthog-core.js:
```diff
@@ -135,6 +137,24 @@ var create_mplib = function (token, config, name) {
+
+    // if target is not defined, we called init after the lib already
+    // loaded, so there won't be an array of things to execute
+    if (!_.isUndefined(target) && _.isArray(target)) {
```
Moving this code is very likely to break something - it has to do with support for if posthog-js is included multiple times and there are many different entrypoints to the script. I would err on being paranoid here - does this _need_ to be moved?

macobo commented on src/posthog-core.js:
```diff
@@ -135,6 +137,24 @@ var create_mplib = function (token, config, name) {
+
+    if (instance.get_config('advanced_disable_decide')) {
+        // When /decide endpoint is disabled, nonone of the features below are available.
+        return instance
```
This is a breaking change. We advertise `posthog.featureFlags` in our documentation, this will make that code throw if a flag is on. 

Suggestion: just gate the decide call.

paolodamico replied:
Agreed, per discussion on Slack

macobo commented on src/posthog-core.js:
```diff
@@ -144,9 +164,8 @@ var create_mplib = function (token, config, name) {
-    // if any instance on the page has debug = true, we set the
-    // global debug to be true
-    Config.DEBUG = Config.DEBUG || instance.get_config('debug')
+    instance.decide = new Decide(instance)
```
Why do we save this on the instance? I don't think there's a real reason to.

macobo reviewed (commented):
Added some notes. Tests seem to be failing as well.

paolodamico commented:
I've addressed all your feedback @macobo, thanks for the thorough review!

I've spent quite a bit of time trying to track down why the testcafe tests are not passing but haven't gotten very far. What's worse is that sometimes the Chrome tests pass, sometimes some of them, sometimes they don't. The only suspect I have is that the /decide call is now being treated async whereas before some condition in the way we run tests was causing it to run synchronously. Can we have a quick sync session to try to figure this out together @macobo ?

macobo commented:
There are some issues with ingestion not working as it did when the e2e test was added - not on this PR. Will be investigating and fixing as soon as cycles free up!

paolodamico commented:
+1 from a user wanting to use the toolbar while having autocapture off
